homeassistant/components/flexmeasures/config_flow.py

```diff
@@ -0,0 +1,239 @@
+        ): str,
+        vol.Required(
+            "username", description={"suggested_value": "toy-user@flexmeasures.io"}
+        ): str,
+        vol.Required("password"): str,
+        vol.Required("power_sensor", description={"suggested_value": 1}): int,
+        vol.Required(
+            "consumption_price_sensor", description={"suggested_value": 2}
+        ): int,
+        vol.Required(
+            "production_price_sensor", description={"suggested_value": 2}
+        ): int,
+        vol.Required("soc_sensor", description={"suggested_value": 4}): int,
+        vol.Required("rm_discharge_sensor", description={"suggested_value": 5}): int,
+        vol.Required(
+            "schedule_duration", description={"suggested_value": "PT24H"}
+        ): str,
+        vol.Required("soc_unit", description={"suggested_value": "kWh"}): str,
+        vol.Required("soc_min", description={"suggested_value": 0.001}): float,
+        vol.Required("soc_max", description={"suggested_value": 0.002}): float,
+    }
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    host, ssl = get_host_and_ssl_from_url(data["url"])
+
+    # Currently used here solely for config validation (i.e. not returned to be stored in the config entry)
+    try:
+        client = FlexMeasuresClient(
+            session=async_get_clientsession(hass),
+            host=host,
+            email=data["username"],
+            password=data["password"],
+            ssl=ssl,
+        )
+    except Exception as exception:
+        raise CannotConnect(exception) from exception
+    try:
+        await client.get_access_token()
+    except Exception as exception:
+        raise InvalidAuth(exception) from exception
+
+    # Return info that you want to store in the config entry.
+    return {"title": "FlexMeasures"}
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Config flow for FlexMeasures integration."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+
+        # Support a single FlexMeasures configuration only
+        await self.async_set_unique_id(DOMAIN)
+        self._abort_if_unique_id_configured()
+
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
+        errors = {}
+
+        try:
+            info = await validate_input(self.hass, user_input)
+        except CannotConnect as exception:
+            errors["base"] = "cannot_connect"
+
+            for field in ("url", "email", "password"):
+                if field in str(exception):
+                    errors[field] = str(exception)
```

Comment:
Instead of this string matching, raise a subclass of `ValueError` in `FlexMeasuresClient` to indicate what failed to validate.